ibis/backends/base/sql/glot/datatypes.py

```diff
@@ -0,0 +1,313 @@
+
+    default_temporal_scale: int | None = None
+    """Default temporal scale when not specified."""
+
+    default_interval_precision: str | None = None
+    """Default interval precision when not specified."""
+
+    unknown_type_strings: dict[str, dt.DataType] = {}
+    """String to ibis datatype mapping to use when converting unknown types."""
+
+    @classmethod
+    def to_ibis(cls, typ: sge.DataType, nullable: bool | None = None) -> dt.DataType:
+        """Convert a sqlglot type to an ibis type."""
+        typecode = typ.this
+
+        if method := getattr(cls, f"_to_ibis_{typecode.name}", None):
+            dtype = method(*typ.expressions)
+        else:
+            dtype = _from_sqlglot_types[typecode](nullable=cls.default_nullable)
+
+        if nullable is not None:
+            return dtype.copy(nullable=nullable)
+        else:
+            return dtype
+
+    @classmethod
+    def from_ibis(cls, dtype: dt.DataType) -> sge.DataType:
+        """Convert a sqlglot type to an ibis type."""
+
+        if method := getattr(cls, f"_from_ibis_{dtype.name}", None):
+            return method(dtype)
+        else:
+            return sge.DataType(this=_to_sqlglot_types[type(dtype)])
+
+    @classmethod
+    def from_string(cls, text: str, nullable: bool | None = None) -> dt.DataType:
+        if dtype := cls.unknown_type_strings.get(text):
+            return dtype
+
+        sgtype = sg.parse_one(text, into=sge.DataType, read=cls.dialect)
+        return cls.to_ibis(sgtype, nullable=nullable)
+
+    @classmethod
+    def to_string(cls, dtype: dt.DataType) -> str:
+        return cls.from_ibis(dtype).sql(dialect=cls.dialect)
+
+    @classmethod
+    def _to_ibis_ARRAY(cls, value_type: sge.DataType) -> dt.Array:
+        return dt.Array(cls.to_ibis(value_type), nullable=cls.default_nullable)
+
+    @classmethod
+    def _to_ibis_MAP(cls, key_type: sge.DataType, value_type: sge.DataType) -> dt.Map:
+        return dt.Map(
+            cls.to_ibis(key_type),
+            cls.to_ibis(value_type),
+            nullable=cls.default_nullable,
+        )
+
+    @classmethod
+    def _to_ibis_STRUCT(cls, *fields: sge.ColumnDef) -> dt.Struct:
+        types = {}
+        for i, field in enumerate(fields):
+            if isinstance(field, sge.ColumnDef):
+                types[field.name] = cls.to_ibis(field.args["kind"])
+            else:
+                types[f"f{i:d}"] = cls.from_string(str(field))
+        return dt.Struct(types, nullable=cls.default_nullable)
+
+    @classmethod
+    def _to_ibis_TIMESTAMP(
+        cls,
+        scale: sge.DataTypeSize | None = None,
+        timezone: sge.DataTypeSize | None = None,
+    ) -> dt.Timestamp:
+        return dt.Timestamp(
+            timezone=None if timezone is None else timezone.this.this,
+            scale=cls.default_temporal_scale if scale is None else int(scale.this.this),
+            nullable=cls.default_nullable,
+        )
+
+    @classmethod
+    def _to_ibis_TIMESTAMPTZ(
```

Comment:
This naming is confusing to me because it looks like it's a method that's converting to an ibis `TIMESTAMPTZ`, which isn't anything that exists.

How about `_from_sqlglot_TIMESTAMPTZ` or `_from_sg_TIMESTAMPTZ`?

Reply:
Yeah, we have kinda 3 options here:

The currently used:
- `_from_ibis_<type>` which is a specialization of `from_ibis()`; type is an ibis dtype
- `_to_ibis_<type>` which is a specialization of `to_ibis()`; type is a sqlglot type

Your proposal:
- `_from_sqlglot_<type>` where type is a sqlglot type
- `_from_ibis_<type>` where type is an ibis dtype

Third:
- `_from_ibis_<type>` where type is an ibis dtype and aligns with `from_ibis()`
- `_<type>_to_ibis` where type is a sqlglot type and it aligns with `to_ibis()`

I'm fine with either one, which is your preference?

Reply:
As you've written this it isn't using an ibis type, it's using a sqlglot type: e.g., `TIMESTAMPTZ` isn't an ibis type, it's from sqlglot.

Reply:
I prefer whatever options do not put the word ibis next to sqlglot type names.

Reply:
Your proposal sounds good to me, going to rename them once I funalize the PR.

Reply:
Updated the naming scheme.